Appveyor failures for persistent 4.2.2 (test failures on 64-bit platforms)

AFAICT the latest persistent release has no Windows wheels on PyPI because https://ci.appveyor.com/project/mgedmin/persistent/build/1.0.31 is all red.

All 64-bit tests fail with

```
======================================================================
FAIL: test_hash_equal (persistent.tests.test_timestamp.PyAndCComparisonTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\projects\persistent\persistent\tests\test_timestamp.py", line 266, in test_hash_equal
    self.assertEqual(hash(c), hash(py))
AssertionError: -1419374591 != 1979033151
======================================================================
FAIL: test_hash_equal_constants (persistent.tests.test_timestamp.PyAndCComparisonTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\projects\persistent\persistent\tests\test_timestamp.py", line 334, in test_hash_equal_constants
    self.assertEqual(c.__hash__(), 1000006000001)
AssertionError: -721379967 != 1000006000001L
----------------------------------------------------------------------
```
(except on Python 3 the 1st error is `AssertionError: -1419374591 != -3850693964765720575`).

I can't figure out why the 32-bit tests failed (an error from `twine upload`?).

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -5,7 +5,7 @@
 ------------------
 
 - Drop use of ``ctypes`` for determining maximum integer size, to increase
-  pure-Python compatibility.
+  pure-Python compatibility. See https://github.com/zopefoundation/persistent/pull/31
 
 - Ensure that ``__slots__`` attributes are cleared when a persistent
   object is ghostified.  (This excluses classes that override

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,6 +1,12 @@
 ``persistent`` Changelog
 ========================
 
+4.2.3 (unreleased)
+------------------
+
+- Fix the hashcode of Python ``TimeStamp`` objects on 64-bit Python on
+  Windows. See https://github.com/zopefoundation/persistent/pull/55
+
 4.2.2 (2016-11-29)
 ------------------
 

--- a/persistent/tests/test_timestamp.py
+++ b/persistent/tests/test_timestamp.py
@@ -280,26 +280,13 @@ class PyAndCComparisonTests(unittest.TestCase):
 
         is_32_bit_hash = orig_maxint == MAX_32_BITS
 
-        orig_c_long = None
-        c_int64 = None
-        c_int32 = None
-        if hasattr(MUT, 'c_long'):
-            import ctypes
-            orig_c_long = MUT.c_long
-            c_int32 = ctypes.c_int32
-            c_int64 = ctypes.c_int64
-            # win32, even on 64-bit long, has funny sizes
-            is_32_bit_hash = c_int32 == ctypes.c_long
-
         try:
             MUT._MAXINT = MAX_32_BITS
-            MUT.c_long = c_int32
 
             py = self._makePy(*self.now_ts_args)
             self.assertEqual(hash(py), bit_32_hash)
 
             MUT._MAXINT = int(2 ** 63 - 1)
-            MUT.c_long = c_int64
             # call __hash__ directly to avoid interpreter truncation
             # in hash() on 32-bit platforms
             if not self._is_jython:
@@ -316,10 +303,6 @@ class PyAndCComparisonTests(unittest.TestCase):
                     384009219096809580920179179233996861765753210540033)
         finally:
             MUT._MAXINT = orig_maxint
-            if orig_c_long is not None:
-                MUT.c_long = orig_c_long
-            else:
-                del MUT.c_long
 
         # These are *usually* aliases, but aren't required
         # to be (and aren't under Jython 2.7).
@@ -334,7 +317,7 @@ class PyAndCComparisonTests(unittest.TestCase):
         import persistent.timestamp as MUT
         # We get 32-bit hash values on 32-bit platforms, or on the JVM
         # OR on Windows (whether compiled in 64 or 32-bit mode)
-        is_32_bit = MUT._MAXINT == (2**31 - 1) or self._is_jython or sys.platform == 'win32'
+        is_32_bit = MUT._MAXINT == (2**31 - 1) or self._is_jython
 
         c, py = self._make_C_and_Py(b'\x00\x00\x00\x00\x00\x00\x00\x00')
         self.assertEqual(hash(c), 8)

--- a/persistent/tests/test_timestamp.py
+++ b/persistent/tests/test_timestamp.py
@@ -280,13 +280,26 @@ class PyAndCComparisonTests(unittest.TestCase):
 
         is_32_bit_hash = orig_maxint == MAX_32_BITS
 
+        orig_c_long = None
+        c_int64 = None
+        c_int32 = None
+        if hasattr(MUT, 'c_long'):
+            import ctypes
+            orig_c_long = MUT.c_long
+            c_int32 = ctypes.c_int32
+            c_int64 = ctypes.c_int64
+            # win32, even on 64-bit long, has funny sizes
+            is_32_bit_hash = c_int32 == ctypes.c_long
+
         try:
             MUT._MAXINT = MAX_32_BITS
+            MUT.c_long = c_int32
 
             py = self._makePy(*self.now_ts_args)
             self.assertEqual(hash(py), bit_32_hash)
 
             MUT._MAXINT = int(2 ** 63 - 1)
+            MUT.c_long = c_int64
             # call __hash__ directly to avoid interpreter truncation
             # in hash() on 32-bit platforms
             if not self._is_jython:
@@ -303,6 +316,10 @@ class PyAndCComparisonTests(unittest.TestCase):
                     384009219096809580920179179233996861765753210540033)
         finally:
             MUT._MAXINT = orig_maxint
+            if orig_c_long is not None:
+                MUT.c_long = orig_c_long
+            else:
+                del MUT.c_long
 
         # These are *usually* aliases, but aren't required
         # to be (and aren't under Jython 2.7).
@@ -317,7 +334,7 @@ class PyAndCComparisonTests(unittest.TestCase):
         import persistent.timestamp as MUT
         # We get 32-bit hash values on 32-bit platforms, or on the JVM
         # OR on Windows (whether compiled in 64 or 32-bit mode)
-        is_32_bit = MUT._MAXINT == (2**31 - 1) or self._is_jython
+        is_32_bit = MUT._MAXINT == (2**31 - 1) or self._is_jython or sys.platform == 'win32'
 
         c, py = self._make_C_and_Py(b'\x00\x00\x00\x00\x00\x00\x00\x00')
         self.assertEqual(hash(c), 8)

--- a/persistent/tests/test_timestamp.py
+++ b/persistent/tests/test_timestamp.py
@@ -12,7 +12,7 @@
 #
 ##############################################################################
 import unittest
-
+import sys
 
 MAX_32_BITS = 2 ** 31 - 1
 MAX_64_BITS = 2 ** 63 - 1
@@ -277,12 +277,29 @@ class PyAndCComparisonTests(unittest.TestCase):
         bit_32_hash = -1419374591
         bit_64_hash = -3850693964765720575
         orig_maxint = MUT._MAXINT
+
+        is_32_bit_hash = orig_maxint == MAX_32_BITS
+
+        orig_c_long = None
+        c_int64 = None
+        c_int32 = None
+        if hasattr(MUT, 'c_long'):
+            import ctypes
+            orig_c_long = MUT.c_long
+            c_int32 = ctypes.c_int32
+            c_int64 = ctypes.c_int64
+            # win32, even on 64-bit long, has funny sizes
+            is_32_bit_hash = c_int32 == ctypes.c_long
+
         try:
             MUT._MAXINT = MAX_32_BITS
+            MUT.c_long = c_int32
+
             py = self._makePy(*self.now_ts_args)
             self.assertEqual(hash(py), bit_32_hash)
 
             MUT._MAXINT = int(2 ** 63 - 1)
+            MUT.c_long = c_int64
             # call __hash__ directly to avoid interpreter truncation
             # in hash() on 32-bit platforms
             if not self._is_jython:
@@ -299,68 +316,73 @@ class PyAndCComparisonTests(unittest.TestCase):
                     384009219096809580920179179233996861765753210540033)
         finally:
             MUT._MAXINT = orig_maxint
+            if orig_c_long is not None:
+                MUT.c_long = orig_c_long
+            else:
+                del MUT.c_long
 
         # These are *usually* aliases, but aren't required
         # to be (and aren't under Jython 2.7).
-        if orig_maxint == MAX_32_BITS:
+        if is_32_bit_hash:
             self.assertEqual(py.__hash__(), bit_32_hash)
-        elif orig_maxint == MAX_64_BITS:
+        else:
             self.assertEqual(py.__hash__(), bit_64_hash)
 
     def test_hash_equal_constants(self):
         # The simple constants make it easier to diagnose
         # a difference in algorithms
         import persistent.timestamp as MUT
-        # We get 32-bit hash values of 32-bit platforms, or on the JVM
-        is_32_bit = MUT._MAXINT == (2**31 - 1) or self._is_jython
+        # We get 32-bit hash values on 32-bit platforms, or on the JVM
+        # OR on Windows (whether compiled in 64 or 32-bit mode)
+        is_32_bit = MUT._MAXINT == (2**31 - 1) or self._is_jython or sys.platform == 'win32'
 
         c, py = self._make_C_and_Py(b'\x00\x00\x00\x00\x00\x00\x00\x00')
-        self.assertEqual(c.__hash__(), 8)
+        self.assertEqual(hash(c), 8)
         self.assertEqual(hash(c), hash(py))
 
         c, py = self._make_C_and_Py(b'\x00\x00\x00\x00\x00\x00\x00\x01')
-        self.assertEqual(c.__hash__(), 9)
+        self.assertEqual(hash(c), 9)
         self.assertEqual(hash(c), hash(py))
 
         c, py = self._make_C_and_Py(b'\x00\x00\x00\x00\x00\x00\x01\x00')
-        self.assertEqual(c.__hash__(), 1000011)
+        self.assertEqual(hash(c), 1000011)
         self.assertEqual(hash(c), hash(py))
 
         # overflow kicks in here on 32-bit platforms
         c, py = self._make_C_and_Py(b'\x00\x00\x00\x00\x00\x01\x00\x00')
         if is_32_bit:
-            self.assertEqual(c.__hash__(), -721379967)
+            self.assertEqual(hash(c), -721379967)
         else:
-            self.assertEqual(c.__hash__(), 1000006000001)
+            self.assertEqual(hash(c), 1000006000001)
         self.assertEqual(hash(c), hash(py))
 
         c, py = self._make_C_and_Py(b'\x00\x00\x00\x00\x01\x00\x00\x00')
         if is_32_bit:
-            self.assertEqual(c.__hash__(), 583896275)
+            self.assertEqual(hash(c), 583896275)
         else:
-            self.assertEqual(c.__hash__(), 1000009000027000019)
+            self.assertEqual(hash(c), 1000009000027000019)
         self.assertEqual(hash(c), hash(py))
 
         # Overflow kicks in at this point on 64-bit platforms
         c, py = self._make_C_and_Py(b'\x00\x00\x00\x01\x00\x00\x00\x00')
         if is_32_bit:
-            self.assertEqual(c.__hash__(), 1525764953)
+            self.assertEqual(hash(c), 1525764953)
         else:
-            self.assertEqual(c.__hash__(), -4442925868394654887)
+            self.assertEqual(hash(c), -4442925868394654887)
         self.assertEqual(hash(c), hash(py))
 
         c, py = self._make_C_and_Py(b'\x00\x00\x01\x00\x00\x00\x00\x00')
         if is_32_bit:
-            self.assertEqual(c.__hash__(), -429739973)
+            self.assertEqual(hash(c), -429739973)
         else:
-            self.assertEqual(c.__hash__(), -3993531167153147845)
+            self.assertEqual(hash(c), -3993531167153147845)
         self.assertEqual(hash(c), hash(py))
 
         c, py = self._make_C_and_Py(b'\x01\x00\x00\x00\x00\x00\x00\x00')
         if is_32_bit:
-            self.assertEqual(c.__hash__(), 263152323)
+            self.assertEqual(hash(c), 263152323)
         else:
-            self.assertEqual(c.__hash__(), -3099646879006235965)
+            self.assertEqual(hash(c), -3099646879006235965)
         self.assertEqual(hash(c), hash(py))
 
     def test_ordering(self):

--- a/persistent/timestamp.py
+++ b/persistent/timestamp.py
@@ -19,7 +19,8 @@ import struct
 import sys
 
 _RAWTYPE = bytes
-_MAXINT = sys.maxsize
+# On win32, long is 4 bytes even on a 64-bit platform.
+_MAXINT = sys.maxsize if not sys.platform == 'win32' else (2 ** 31 - 1)
 
 def _makeOctets(s):
     if sys.version_info < (3,):
@@ -29,19 +30,11 @@ def _makeOctets(s):
 
 _ZERO = _makeOctets('\x00' * 8)
 
-try:
+def _wraparound(x):
     # Make sure to overflow and wraparound just
-    # like the C code does.
-    from ctypes import c_long
-except ImportError: # pragma: no cover
-    # XXX: This is broken on 64-bit windows, where
-    # sizeof(long) != sizeof(Py_ssize_t)
-    # sizeof(long) == 4, sizeof(Py_ssize_t) == 8
-    def _wraparound(x):
-        return int(((x + (_MAXINT + 1)) & ((_MAXINT << 1) + 1)) - (_MAXINT + 1))
-else:
-    def _wraparound(x):
-        return c_long(x).value
+    # like the C code does. This is more compactly expressed
+    # as ctypes.c_long(x).value
+    return int(((x + (_MAXINT + 1)) & ((_MAXINT << 1) + 1)) - (_MAXINT + 1))
 
 class _UTC(datetime.tzinfo):
     def tzname(self):

--- a/persistent/timestamp.py
+++ b/persistent/timestamp.py
@@ -19,8 +19,7 @@ import struct
 import sys
 
 _RAWTYPE = bytes
-# On win32, long is 4 bytes even on a 64-bit platform.
-_MAXINT = sys.maxsize if not sys.platform == 'win32' else (2 ** 31 - 1)
+_MAXINT = sys.maxsize
 
 def _makeOctets(s):
     if sys.version_info < (3,):
@@ -30,11 +29,22 @@ def _makeOctets(s):
 
 _ZERO = _makeOctets('\x00' * 8)
 
-def _wraparound(x):
+try:
     # Make sure to overflow and wraparound just
-    # like the C code does. This is more compactly expressed
-    # as ctypes.c_long(x).value
-    return int(((x + (_MAXINT + 1)) & ((_MAXINT << 1) + 1)) - (_MAXINT + 1))
+    # like the C code does.
+    from ctypes import c_long
+except ImportError: # pragma: no cover
+    # XXX: This is broken on 64-bit windows, where
+    # sizeof(long) != sizeof(Py_ssize_t)
+    # sizeof(long) == 4, sizeof(Py_ssize_t) == 8
+    # It can be fixed by setting _MAXINT = 2 ** 31 - 1 on all
+    # win32 platforms, but then that breaks PyPy3 64 bit for an unknown
+    # reason.
+    def _wraparound(x):
+        return int(((x + (_MAXINT + 1)) & ((_MAXINT << 1) + 1)) - (_MAXINT + 1))
+else:
+    def _wraparound(x):
+        return c_long(x).value
 
 class _UTC(datetime.tzinfo):
     def tzname(self):

--- a/persistent/timestamp.py
+++ b/persistent/timestamp.py
@@ -29,12 +29,19 @@ def _makeOctets(s):
 
 _ZERO = _makeOctets('\x00' * 8)
 
-
-def _wraparound(x):
+try:
     # Make sure to overflow and wraparound just
     # like the C code does.
-    return int(((x + (_MAXINT + 1)) & ((_MAXINT << 1) + 1)) - (_MAXINT + 1))
-
+    from ctypes import c_long
+except ImportError: # pragma: no cover
+    # XXX: This is broken on 64-bit windows, where
+    # sizeof(long) != sizeof(Py_ssize_t)
+    # sizeof(long) == 4, sizeof(Py_ssize_t) == 8
+    def _wraparound(x):
+        return int(((x + (_MAXINT + 1)) & ((_MAXINT << 1) + 1)) - (_MAXINT + 1))
+else:
+    def _wraparound(x):
+        return c_long(x).value
 
 class _UTC(datetime.tzinfo):
     def tzname(self):

